man/man1/wperfmon.1
.TH WPERFMON 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wperfmon \- a performance measurement tool for W
.SH SYNOPSIS
wperfmon
.SH DESCRIPTION
.I Wperfmon
is the new tools to check if everything works OK and how much time is needed
for some kind of operations. The first and probably most important thing is
the null operation which simply send NOP packets to the wserver to calculate
the socket overhead. All other calls really should be slower than that. It
doesn't attempt to give you netto graphic performance numbers, but this may
change in future.
.PP
Be sure that the system is free of load when running wperfmon. It is probably
also a wise idea to let it be the only client running because the time
measurement routines are a bit stupid...
.SH LAST CHANGED
15th. Feb. 1996 by TeSche


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


benchmark/wperfmon.c
/*
 * wperfmon.c, a part of the W Window System
 *
 * Copyright (C) 1994-1997 by Torsten Scherer
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- a performance monitor for W1
 *
 * CHANGES
 * ++eero, 6/98:
 * - Sun and netBSD use now util library getopt.
 * - Uses w_gettime() instead of earlier now() function.
 */

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <sys/time.h>
#include <getopt.h>
#include <Wlib.h>

#define FNAME "fixed"
#define FSIZE 8
#define	REPS 5

#define WIDTH 320
#define HEIGHT 200

static int width, height, rows, columns;


/*
 * something...
 */

static WWIN *win;
static short fw, fh, end;


static void sigalrm(int sig)
{
  end = 1;
}


/*
 * some tests
 */

static void test_null(short reps, long sec)
{
  short i;
  long vorher, l, suml;
  float s, sums;

  printf("testing null operation...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    vorher = w_gettime();
    l = 0;
    end = 0;
    alarm(sec);
    while (!end) {
      l++;
      w_null();
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_plot(short reps, long sec)
{
  short x, y, i;
  long vorher, l, suml;
  float s, sums;

  printf("testing plot...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    x = 0;
    y = 0;
    l = 0;
    end = 0;
    w_setmode(win, M_CLEAR);
    w_pbox(win, 0, 0, width, height);
    w_setmode(win, M_INVERS);
    w_flush();
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      w_plot(win, x, y);
      l++;
      if ((x += 13) >= width) {
	x -= width;
	if (++y == height) {
	  y = 0;
	}
      }
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;

    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_line(short reps, long sec)
{
  short i, j;
  long vorher, l, suml;
  float s, sums;

  printf("testing line...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    l = 0;
    end = 0;
    w_setmode(win, M_CLEAR);
    w_pbox(win, 0, 0, width, height);
    w_setmode(win, M_INVERS);
    w_flush();
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      for (j=0; j<width; j++)
	w_line(win, j, 0, width-1-j, height-1);
      for (j=0; j<height; j++)
	w_line(win, 0, j, width-1, height-1-j);
      l += width + height;
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_vline(short reps, long sec)
{
  short x, i;
  long vorher, l, suml;
  float s, sums;

  printf("testing vline...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    x = 0;
    l = 0;
    end = 0;
    w_setmode(win, M_CLEAR);
    w_pbox(win, 0, 0, width, height);
    w_setmode(win, M_INVERS);
    w_flush();
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      w_vline(win, x, 0, height-1);
      l++;
      if (++x == width) {
	x = 0;
      }
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_hline(short reps, long sec)
{
  short y, i;
  long vorher, l, suml;
  float s, sums;

  printf("testing hline...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    y = 0;
    l = 0;
    end = 0;
    w_setmode(win, M_CLEAR);
    w_pbox(win, 0, 0, width, height);
    w_setmode(win, M_INVERS);
    w_flush();
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      w_hline(win, 0, y, width-1);
      l++;
      if (++y == height) {
	y = 0;
      }
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


#define POLYPOINTS 9

static void test_ppoly(short reps, long sec)
{
  short i;
  long vorher, l, suml;
  float s, sums;
  short points[POLYPOINTS<<1] = {0, height>>1, width>>2, 0, width>>1, height>>2, (width>>2)*3, 0, width-1, height>>1, (width>>2)*3, height-1, width>>1, (height>>2)*3, width>>2, height-1, 0, height>>1};

  printf("testing ppoly...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    l = 0;
    end = 0;
    w_setmode(win, M_CLEAR);
    w_pbox(win, 0, 0, width, height);
    w_setmode(win, M_INVERS);
    w_flush();

    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      w_ppoly(win, POLYPOINTS, points);
      l++;
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_circle(short reps, long sec)
{
  short x, y, i;
  long vorher, l, suml;
  float s, sums;

  printf("testing circle...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    x = 50;
    y = 50;
    l = 0;
    end = 0;
    w_setmode(win, M_CLEAR);
    w_pbox(win, 0, 0, width, height);
    w_setmode(win, M_INVERS);
    w_flush();
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      w_circle(win, x, y, 49);
      l++;
      if (++x >= width-50) {
	x = 50;
	if (++y >= height-50) {
	  y = 50;
	}
      }
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_pcircle(short reps, long sec)
{
  short x, y, i;
  long vorher, l, suml;
  float s, sums;

  printf("testing pcircle...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    x = 50;
    y = 50;
    l = 0;
    end = 0;
    w_setmode(win, M_CLEAR);
    w_pbox(win, 0, 0, width, height);
    w_setmode(win, M_INVERS);
    w_flush();
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      w_pcircle(win, x, y, 49);
      l++;
      if (++x >= width-50) {
	x = 50;
	if (++y >= height-50) {
	  y = 50;
	}
      }
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_bitblk(short reps, long sec)
{
  short i, j;
  long vorher, l, suml;
  float s, sums;

  printf("testing bitblk...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    w_setmode(win, M_CLEAR);
    w_pbox(win, 0, 0, width, height);
    w_setmode(win, M_INVERS);
    for (j=0; j<width; j++)
      w_line(win, j, 0, width-1-j, height-1);
    for (j=0; j<height; j++)
      w_line(win, 0, j, width-1, height-1-j);
    w_flush();
    l = 0;
    end = 0;
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      w_bitblk(win, 1, 0, width-1, height, 0, 0);
      l++;
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_scroll(short reps, long sec)
{
  short i, j;
  long vorher, l, suml;
  float s, sums;

  printf("testing vscroll...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    w_setmode(win, M_CLEAR);
    w_pbox(win, 0, 0, width, height);
    w_setmode(win, M_INVERS);
    for (j=0; j<width; j++)
      w_line(win, j, 0, width-1-j, height-1);
    for (j=0; j<height; j++)
      w_line(win, 0, j, width-1, height-1-j);
    w_flush();
    l = 0;
    end = 0;
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      w_vscroll(win, 0, 1, width, height-1, 0);
      l++;
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_printc(short reps, long sec)
{
  short x, y, i;
  long vorher, l, suml;
  float s, sums;

  printf("testing character output...\n");

  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    x = 0;
    y = 0;
    l = 0;
    end = 0;
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      w_printchar(win, x, y, (l++ % 63) + 32);
      if ((x += fw) == width) {
	x = 0;
	if ((y += fh) == height) {
	  y = 0;
	}
      }
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


static void test_prints(short reps, long sec)
{
  short y, i, j;
  long vorher, l, suml;
  float s, sums;
  char buf[65];

  printf("testing string output...\n");

  buf[64] = 0;
  suml = 0;
  sums = 0;
  for (i=0; i<reps; i++) {
    y = 0;
    l = 0;
    end = 0;
    vorher = w_gettime();
    alarm(sec);
    while (!end) {
      for (j=0; j<64; j++) {
	buf[j] = (l++ % 63) + 32;
      }
      w_printstring(win, 0, y, buf);
      if ((y += fh) == height) {
	y = 0;
      }
    }
    w_test(win, 0, 0);
    s = (w_gettime() - vorher) / 1000.0;
    printf("%li ops in %f seconds -> %f ops/sec\n", l, s, l / s);
    suml += l;
    sums += s;
  }

  printf("average: %f secs/op, %f ops/sec\n", sums / suml, suml / sums);
}


/*
 *
 */

static int init (void)
{
  WSERVER *wserver;
  WFONT *wfont;

  struct sigaction sa;

  sa.sa_handler = sigalrm;
  sigemptyset(&sa.sa_mask);
#if defined(linux)
  sa.sa_flags = SA_RESTART;
#else
  sa.sa_flags = 0;
#endif
  sigaction(SIGALRM, &sa, NULL);

  if (!(wserver = w_init())) {
    fprintf(stderr, "error: can't connect to wserver\n");
    return -1;
  }

  if (!(wfont = w_loadfont(FNAME, FSIZE, 0))) {
    fprintf(stderr, "error: can't load font\n");
    return -1;
  }

  if (wfont->flags & F_PROP) {
    fprintf(stderr, "error: this will not work with a proportional font\n");
    return -1;
  }

  fw = wfont->widths[0];
  fh = wfont->height;
  columns = WIDTH / fw;
  rows = HEIGHT / fh;
  width = columns * fw;
  height = rows * fh;

  if ((wserver->width < width) || (wserver->height < height)) {
    fprintf(stderr, "error: screen size (%ix%i) too small (wanted %ix%i)\n",
	    wserver->width, wserver->height, width, height);
    return -1;
  }

  if (!(win = w_create(width, height, W_TOP | W_NOBORDER))) {
    fprintf(stderr, "error: can't create window\n");
    return -1;
  }

  if (w_open(win, 0, 0) < 0) {
    w_delete(win);
    fprintf(stderr, "error: can't open window\n");
    return -1;
  }

  printf("This is WPERFMON, running on W%iR%i",
	 wserver->vmaj, wserver->vmin);
  if (wserver->pl)
    printf (" PL%i", wserver->pl);
  printf("\nscreen: %ix%i, %i colors\n",
	 wserver->width, wserver->height, 1 << wserver->planes);
  printf("window: %ix%i\n", width, height);
  printf("font: %ix%i (%s/%d)\n", fw, fh, FNAME, FSIZE);

  w_setfont(win, wfont);

  return 0;
}


/*
 * guess what...
 */

typedef struct {
  const char *name;
  int doit;
  void (*func)(short, long);
} TEST;

static TEST tests[] = {
  {"null", 0, test_null},
  {"plot", 0, test_plot},
  {"line", 0, test_line},
  {"vline", 0, test_vline},
  {"hline", 0, test_hline},
  {"ppoly", 0, test_ppoly},
  {"circle", 0, test_circle},
  {"pcircle", 0, test_pcircle},
  {"bitblk", 0, test_bitblk},
  {"scroll", 0, test_scroll},
  {"printc", 0, test_printc},
  {"prints", 0, test_prints},
  {NULL, 0, NULL}
};

static void usage (char *prgname)
{
  const TEST *test;

  fprintf(stderr, "usage: %s [<test>]\n", prgname);
  fprintf(stderr, "       where <test> may be any combination of:\n");

  for (test=tests; test->name; test++) {
    fprintf(stderr, "       --%s\n", test->name);
  }

  exit (-1);
}


static const struct option options[] = {
  {"null", 0, NULL, 0},
  {"plot", 0, NULL, 1},
  {"line", 0, NULL, 2},
  {"vline", 0, NULL, 3},
  {"hline", 0, NULL, 4},
  {"ppoly", 0, NULL, 5},
  {"circle", 0, NULL, 6},
  {"pcircle", 0, NULL, 7},
  {"bitblk", 0, NULL, 8},
  {"scroll", 0, NULL, 9},
  {"printc", 0, NULL, 10},
  {"prints", 0, NULL, 11},
  {"all", 0, NULL, 127},
  {NULL, 0, NULL, 0}
};


int main (int argc, char *argv[])
{
  int ret;
  TEST *test;

  /* parse arguments
   */
  while ((ret = getopt_long(argc, argv, "", options, NULL)) != -1) {
    switch(ret) {
      case 0:
      case 1:
      case 2:
      case 3:
      case 4:
      case 5:
      case 6:
      case 7:
      case 8:
      case 9:
      case 10:
      case 11:
        tests[ret].doit = 1;
	break;

      case 127:
	for (test=tests; test->name; test++)
	  test->doit = 1;
	break;

      case ':':
	printf("missing argument for option '%s'\n", argv[optind]);
	usage(argv[0]);

      default:
	usage(argv[0]);
    }
  }

  /* no additional arguments supported so far
   */
  if ((optind < argc) || (argc < 2)) {
    usage(argv[0]);
  }

  if (init ())
    return -1;

  for (test=tests; test->name; test++) {
    if (test->doit) {
      (*test->func)(REPS, 10);
    }
  }

  return 0;
}
